### DISCUSSION TITLE
GPT-4 Vision-based footage analyst

### DISCUSSION THREAD

**Discussion by mhaowork** [blakeblackshear/frigate#8675] [Show and tell]
Hi 👋! I have been a Frigate user for a while and have always been impressed by Frigate's versatility. After switching from Nest to Frigate, I tried to replicate the package delivery notification functionality of the older camera system. This month, with the release of the GPT-4 Vision API, I was able to take my experimentation to the next level to allow a higher level of contextual understanding.

Here is a prototype called AmbleGPT https://github.com/mhaowork/amblegpt I put together quickly over the past couple weeks. Feel free to give it a shot!

**Feedback and suggestions are welcome!**

### Summary

AmbleGPT is activated by a Frigate event via MQTT and analyzes the event clip using the OpenAI GPT-4 Vision API. It returns an easy-to-understand, context-rich summary. AmbleGPT then publishes this summary text in an MQTT message. The message can be received by the Home Assistant Frigate Notification Automation and sent to a user via iOS/Android notifications.

### Demo

![Notification: Two people crossing street](https://github.com/mhaowork/amblegpt/blob/main/assets/notif_person_on_street.jpeg?raw=true)

More video examples:

| Video        | GPT Summary    |       
| ------------- |:-------------:|
| ![](https://github.com/mhaowork/amblegpt/blob/main/assets/ups_delivery.gif?raw=true)      | A delivery man, approximately 35 years old, approached the door and placed a package down. He briefly interacted with a mobile device before leaving the scene. |
| ![](https://github.com/mhaowork/amblegpt/blob/main/assets/female_waiting_at_door_480p.gif?raw=true)      | A female, approximately 30 years old and 1.65 meters tall, is seen approaching and standing at the front door, looking down momentarily and then preparing to interact with the person who might open the door |
| ![](https://github.com/mhaowork/amblegpt/blob/main/assets/usps_delivery_480p.gif?raw=true)      | A postal worker (in a blue uniform) was seen exiting a delivery vehicle and walking off-screen, presumably to deliver mail or a package. |
| ![](https://github.com/mhaowork/amblegpt/blob/main/assets/two_persons_walking_street.gif?raw=true)         | A male and a female, appearing to be in their 30s, are seen crossing the street from the left to the right. They walk side by side and are visible for a total of 18 seconds.|

More details are in https://github.com/mhaowork/amblegpt

**Comment by kirill578** [comment]
This is a cool project that can defenity help with the false positives. here are a couple of thoughts I have:

1. To accurately determine if someone is acting suspicious, one idea could be to chain prompts. Start with a limited set of frames to make an initial evaluation. If the results are unclear, gradually introduce more frames until you can figure out what's happening in the scene. This approach might help in reducing costs while maintaining effectiveness.

2. Another idea is to provide additional context through the prompt. For example, if someone is opening a door in the middle of the night and then later spotted outside, it's less suspicious compared to the reverse order. By incorporating this kind of information, we can potentially enhance the accuracy of the system.

What are your thoughts?

**Reply by mhaowork** [reply]
Love your ideas @kirill578 ! 

1. This is totally something I am excited about! It can also introduce some level of interactivity and allow users to ask questions to the GPT. I will add it to the future work.

2. Nice point too! More parameters can be to the prompt such as time to better inform the AI. This one is actually easy to implement. I can add it quickly tomorrow / feel free to make PRs as well!

**Reply by mhaowork** [reply]
I added event time to the prompt and the result is pretty interesting! 

With a same following video clip, GPT would give different results based on the event time:

<img src='https://github.com/mhaowork/amblegpt/blob/main/assets/female_waiting_at_door_480p.gif?raw=true' width="250"  >

local_time_str = "2021-09-01 03:15:00"
"summary": "A female appears to be looking for something or someone, shown in a sequence of images taken at night. She seems to be initially looking at a distance and then directly at the camera, with an unclear purpose, which may be considered unusual at this time. Her behavior might be viewed as suspicious due to the late hour and lack of clear intent."

vs

local_time_str = "2021-09-01 19:15:00"
"summary": "A female appears at the doorstep, seems to look for something or wait momentarily before leaving the view of the camera. This series of events occurs at night time, captured by an infrared camera, so the specific details such as facial features are indiscernible. Her behavior does not appear overtly suspicious as she does not attempt to force entry or take anything."

**Comment by mhaowork** [comment]
Added custom prompt feature to allow the GPT to understand each camera's environment. 

For example, 
```
  Front Door:
    # The custom prompt is optional and is used to provide more context for GPT to better understand the footage.
    # This will be inserted into the prompt before sending it to GPT.
    custom_prompt: >
      The camera that took the footage is a doorbell camera outside the front door of the house.
      The door is on the left side of the video frame and the exit is on the right side.
  Outside:
    custom_prompt: >
      The camera that took the footage is mounted on the frame of the front window of the house facing the street.
      The street allow guest parking and it is normal to see cars parked on the street.
      There might also be people walking (with their dogs) on the sidewalk before 9pm.
```
